core/src/sigverify.rs

```diff
@@ -118,24 +181,27 @@ pub fn generate_offsets(batches: &[Packets], recycler: &Recycler<TxOffset>) -> R
         p.packets.iter().for_each(|packet| {
             let current_offset = current_packet as u32 * size_of::<Packet>() as u32;
 
-            let (sig_len, sig_start, msg_start_offset, pubkey_offset) =
-                get_packet_offsets(packet, current_offset);
-            let mut pubkey_offset = pubkey_offset;
+            let packet_offsets = get_packet_offsets(packet, current_offset);
 
-            sig_lens.push(sig_len);
+            sig_lens.push(packet_offsets.sig_len);
 
-            trace!("pubkey_offset: {}", pubkey_offset);
-            let mut sig_offset = sig_start;
-            for _ in 0..sig_len {
-                signature_offsets.push(sig_offset);
-                sig_offset += size_of::<Signature>() as u32;
+            trace!("pubkey_offset: {}", packet_offsets.pubkey_start);
+            if packet_offsets.sig_len > 0 {
```

Comment:
Nice catch! Well, the nuance of this part is rather subtle, so no wonder you have such a question....

Yeah, that's correct. From a functional viewpoint, this `if` is redundant.

The reason I added this `if` is to avoid unneeded memory reads for (hypothetical, noted later) performance reasons. I presume this function is in a tight hot loop, potentially being a bottleneck, positioned at immediately before feeding the offsets to the GPU.

So, I wrote to skip unneeded reads of `packet_offsets.pubkey_start` and `packet_offsets.sig_start` if `sig_len == 0`.
Of course, CPU's branch prediction/prefeching and/or `rustc`/llvm might be smart enough to make this redundant coding render meaningless. Alfter all, this is a classic correct/error conditioning, resulting in very likely/very unlikly dichotomy in a run-time wise..

If that's true, I'm fine to remove the `if`. If it's not, maybe should I comment the above reasoning?

Note: Sorry but this is not based on any measured benchmark, so with little substance. I'm stating this just as a educated guess... :p

Reply:
Heh, out of curiosity, I investigated this a bit. Happily, I was wrong.

`ructc` and LLVM are smart enough to emit identical x86-64 machine codes for both cases of `let`s both inside and outside `if`!: https://rust.godbolt.org/z/Oko61h

So, the above comment is totally moot.

Then, I guess the remaining justification for putting `let`s under the `if` is for code readability with relevant variables scoped as small as possible to reduce any human's cognitive load, which is debatable due to subjectivity though. However, I think I moved `pubkey_offset` here for the better from a rather distant location from its use.

Reply:
@ryoqun , thank you for looking into that!
Let's remove the `if` statement for fewer lines of code, but definitely leave your reorg.
I have a couple cast nits in your test below, since you're iterating again. Then I'd love to get this merged, as I want to build on it :)